File: src/main/java/com/destino/projeto_destino/config/SecurityConfiguration.java
package com.destino.projeto_destino.config;

import com.destino.projeto_destino.repository.usuario.UsuarioRepository;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.access.hierarchicalroles.RoleHierarchy;
import org.springframework.security.access.hierarchicalroles.RoleHierarchyImpl;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.config.Customizer;
import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration;
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;

import java.util.Arrays;
import java.util.List;

@Configuration
@EnableWebSecurity
@EnableMethodSecurity(prePostEnabled = true)
public class SecurityConfiguration {

    private final JwtAuthenticationFilter jwtAuthenticationFilter;
    private final UsuarioRepository userRepository;

    public SecurityConfiguration(
            JwtAuthenticationFilter jwtAuthenticationFilter,
            UsuarioRepository userRepository
    ) {
        this.jwtAuthenticationFilter = jwtAuthenticationFilter;
        this.userRepository = userRepository;
    }

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
        return http
                .cors(Customizer.withDefaults())
                .csrf(AbstractHttpConfigurer::disable)
                .sessionManagement(session -> session.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
                .formLogin(AbstractHttpConfigurer::disable)
                .addFilterBefore(jwtAuthenticationFilter, BasicAuthenticationFilter.class)
                .authorizeHttpRequests(auth -> auth
                        // Swagger
                        .requestMatchers("/swagger-ui/**", "/v3/api-docs/**").permitAll()

                        // Endpoints Públicos
                        .requestMatchers(
                                "/api/compra/**",
                                "/api/test/publico/**",
                                "/api/publico/pacote/**"
                        ).permitAll()

                        // AuthController (Rotas públicas de autenticação)
                        .requestMatchers(
                                "/api/auth/cadastrar",
                                "/api/auth/entrar",
                                "/api/auth/renovar-token",
                                "/api/auth/sair"
                        ).permitAll()

                        .requestMatchers(
                                "/api/compra/**"
                        ).hasRole("USUARIO")

                        // Rotas para funcionários
                        .requestMatchers(
                                "/api/pacote/**",
                                "/api/pacote/{id}",
                                "/api/usuario/**",
                                "/api/pacote/agrupado-admin",
                                "/api/hotel/**",
                                "/api/transporte/**",
                                "/api/dashboard/**",
                                "/api/pacote-foto/**"
                        ).hasRole("FUNCIONARIO")

                        // Endpoints Protegidos
                        .requestMatchers("/api/usuario/**").authenticated()
                        .requestMatchers("/api/test/privado/**").authenticated()

                        .anyRequest().authenticated()
                ).build();
    }

    @Bean
    public RoleHierarchy roleHierarchy() {
        return RoleHierarchyImpl.fromHierarchy(
                "ROLE_ADMINISTRADOR > ROLE_FUNCIONARIO \n" +
                        "ROLE_FUNCIONARIO > ROLE_USUARIO"
        );
    }

    @Bean
    public CorsConfigurationSource corsConfigurationSource() {
        CorsConfiguration configuration = new CorsConfiguration();
        configuration.setAllowedOrigins(List.of("https://localhost:5173"));
        configuration.setAllowedMethods(Arrays.asList("GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"));
        configuration.setAllowedHeaders(List.of("*"));
        configuration.setAllowCredentials(true);

        UrlBasedCorsConfigurationSource source = new UrlBasedCorsConfigurationSource();
        source.registerCorsConfiguration("/**", configuration); // Aplica a todas as URLs

        return source;
    }

    @Bean
    UserDetailsService userDetailsService() {
        return email -> userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("Email não encontrado: " + email));
    }

    @Bean
    BCryptPasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder(12);
    }

    @Bean
    public AuthenticationManager authenticationManager(AuthenticationConfiguration config) throws Exception {
        return config.getAuthenticationManager();
    }
}

File: src/main/kotlin/com/fatec/destino/config/SecurityConfiguration.kt
package com.fatec.destino.config

import com.fatec.destino.repository.usuario.UsuarioRepository
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.access.hierarchicalroles.RoleHierarchy
import org.springframework.security.access.hierarchicalroles.RoleHierarchyImpl
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.invoke
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import org.springframework.web.cors.UrlBasedCorsConfigurationSource

@Configuration
@EnableWebSecurity
@EnableMethodSecurity(prePostEnabled = true)
class SecurityConfiguration(
    private val jwtAuthenticationFilter: JwtAuthenticationFilter,
    private val userRepository: UsuarioRepository
) {

    @Bean
    fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
        http.authorizeHttpRequests {
            it.requestMatchers("/v3/api-docs/**", "/swagger-ui/**").permitAll()
        }
        http {
            cors { configurationSource = corsConfigurationSource() }
            csrf { disable() }
            sessionManagement { sessionCreationPolicy = SessionCreationPolicy.STATELESS }
            formLogin { disable() }

            // Adiciona o filtro JWT antes do filtro de autenticação básica
            addFilterBefore<BasicAuthenticationFilter>(jwtAuthenticationFilter)

            authorizeHttpRequests {
                // 1. Swagger (Sempre primeiro)
                authorize("/v3/api-docs/**", permitAll)
                authorize("/v3/api-docs.yaml", permitAll)
                authorize("/swagger-ui/**", permitAll)
                authorize("/swagger-ui.html", permitAll)
                authorize("/swagger-resources/**", permitAll)
                authorize("/webjars/**", permitAll)

                // 2. AuthController (Rotas públicas de autenticação)
                authorize("/api/auth/cadastrar", permitAll)
                authorize("/api/auth/entrar", permitAll)
                authorize("/api/auth/renovar-token", permitAll)
                authorize("/api/auth/sair", permitAll)

                // 3. Endpoints Públicos de Negócio
                authorize("/api/test/publico/**", permitAll)
                authorize("/api/publico/pacote/**", permitAll)

                // Nota: Removi /api/compra/** do permitAll pois você a definiu
                // logo abaixo como restrita ao ROLE_USUARIO

                // 4. Roles Específicas
                authorize("/api/compra/**", hasRole("USUARIO"))

                // 5. Rotas para funcionários (Hierarquia)
                listOf(
                    "/api/pacote/**",
                    "/api/pacote/{id}",
                    "/api/usuario/**",
                    "/api/pacote/agrupado-admin",
                    "/api/hotel/**",
                    "/api/transporte/**",
                    "/api/dashboard/**",
                    "/api/pacote-foto/**"
                ).forEach { path -> authorize(path, hasRole("FUNCIONARIO")) }

                // 6. Endpoints Protegidos Genéricos
                authorize("/api/usuario/**", authenticated)
                authorize("/api/test/privado/**", authenticated)

                // 7. Qualquer outra requisição
                authorize(anyRequest, authenticated)
            }
        }
        return http.build()
    }

    @Bean
    fun roleHierarchy(): RoleHierarchy {
        return RoleHierarchyImpl.fromHierarchy(
            """
            ROLE_ADMINISTRADOR > ROLE_FUNCIONARIO
            ROLE_FUNCIONARIO > ROLE_USUARIO
            """.trimIndent()
        )
    }

    @Bean
    fun corsConfigurationSource(): CorsConfigurationSource {
        val configuration = CorsConfiguration().apply {
            allowedOrigins = listOf("https://localhost:5173")
            allowedMethods = listOf("GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS")
            allowedHeaders = listOf("*")
            allowCredentials = true
        }

        return UrlBasedCorsConfigurationSource().apply {
            registerCorsConfiguration("/**", configuration)
        }
    }

    @Bean
    fun userDetailsService(): UserDetailsService = UserDetailsService { email ->
        userRepository.findByEmail(email) ?: throw RuntimeException("Usuário não encontrado")
    }

    @Bean
    fun passwordEncoder() = BCryptPasswordEncoder(12)

    @Bean
    fun authenticationManager(config: AuthenticationConfiguration): AuthenticationManager {
        return config.authenticationManager
    }
}